### PR TITLE
Add support for Elasticsearch 7

### PR DESCRIPTION
**Pull request by kevingosse**
Contributes to https://github.com/DataDog/dd-trace-dotnet/issues/1584

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -357,6 +368,7 @@ services:
       - MONGO_HOST=mongo_arm64
       - SERVICESTACK_REDIS_HOST=servicestackredis_arm64:6379
       - STACKEXCHANGE_REDIS_HOST=stackexchangeredis_arm64:6379
+      - ELASTICSEARCH7_HOST=elasticsearch7_arm64:9200
```
Comment (andrewlock) [inline docker-compose.yml:371]: You'll need to create the `elasticsearch7_arm64` service too (and add it to StartDependencies.ARM64)

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/IApiCallDetails.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// <copyright file="IApiCallDetails.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+#pragma warning disable SA1600 // Elements should be documented
```
Comment (andrewlock) [inline src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/IApiCallDetails.cs:7]: Cheat 😛

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            var uri = response?.ApiCall?.Uri?.ToString();
+
+            if (uri != null)
+            {
+                var tags = (ElasticsearchTags)state.State;
+
+                if (tags != null)
+                {
+                    tags.Url = uri;
+                }
```
Comment (andrewlock) [inline src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs:72]: Doesn't need to change, just pointing out "modern" syntax
```suggestion
                if (state.State is ElasticsearchTags tags)
                {
                    tags.Url = uri;
                }
```

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            }
+
+            state.Scope.DisposeWithException(exception);
```
Comment (andrewlock) [inline src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs:75]: Scope could be null if integration is disabled
```suggestion
            state.Scope?.DisposeWithException(exception);
```

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_RequestAsync_Integration.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                var tags = (ElasticsearchTags)state.State;
+
+                if (tags != null)
+                {
+                    tags.Url = uri;
+                }
+            }
+
+            state.Scope.DisposeWithException(exception);
```
Comment (andrewlock) [inline src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_RequestAsync_Integration.cs:77]: Scope could be null if integration is disabled
```suggestion
            state.Scope?.DisposeWithException(exception);
```

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+                // commands with sync and async
+                for (var i = 0; i < 2; i++)
+                {
+                    expected.AddRange(new List<string>
+                    {
+                        "Bulk",
+                        "Create",
+                        "Search",
+                        "DeleteByQuery",
```
Comment (andrewlock) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs:51]: Are these the same in all versions of ES (Might be nice to extract them and be explicit if so)

**Review by andrewlock — APPROVED** [review]
Looks good 👍

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs (hunk 4 above)
Comment (kevingosse) [reply src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs:75]: `DisposeWithException` has a null-check integrated: https://github.com/DataDog/dd-trace-dotnet/blob/a72b74fc8e67d8d8a6430628fe8643b3a693d2bc/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AutoInstrumentationExtensions.cs#L16

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs (hunk 6 above)
Comment (kevingosse) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs:51]: They're different sample projects so I don't think there's value in de-duplicating the list of spans

File: docker-compose.yml (hunk 1 above)
Comment (kevingosse) [reply docker-compose.yml:371]: You mean this one? 😄  https://github.com/DataDog/dd-trace-dotnet/blob/master/docker-compose.yml#L9

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/IApiCallDetails.cs (hunk 2 above)
Comment (kevingosse) [reply src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/IApiCallDetails.cs:7]: 😎

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs (hunk 3 above)
Comment (kevingosse) [reply src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs:72]: I always forget this also does a null check

File: docker-compose.yml (hunk 1 above)
Comment (andrewlock) [reply docker-compose.yml:371]: Err.... yeah... exactly like that one. Silly me, thinking we would be using elasticsearch 5 for elasticsearch5 tests 😬

File: src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs (hunk 4 above)
Comment (andrewlock) [reply src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Elasticsearch/V7/Transport_Request_Integration.cs:75]: Oh yeah. I feel like I commented for someone to remove the null check recently 🤦‍♂️

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs (hunk 6 above)
Comment (andrewlock) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch7Tests.cs:51]: Fair enough. I just wondered if there were differences in supported values between projects

**Review by zacharycmontoya — APPROVED** [review]
Looks good!

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will **not change** line coverage
:heavy_check_mark: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will **not change** branch coverage
:no_entry: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will will **increase** complexity by `23`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) | #1760       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9314` / `14757` | `9273` / `14720` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4126` / `6418` | `4134` / `6440` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7558`      | `7581`        |  `23`  :no_entry:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75197/artifacts?artifactName=Code%20Coverage%20Report_75197&api-version=6.0&%24format=zip)
* [For this PR #1760](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75180/artifacts?artifactName=Code%20Coverage%20Report_75180&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) | #1760       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7558`      | `7581`       |  `23` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Logging.LogRateLimiter](https://github.com/DataDog/dd-trace-dotnet/tree/5b54a7965f259629d6a4da155096561772bbf208/tracer/src/Datadog.Trace/Logging/LogRateLimiter.cs) | `0.00%` :heavy_check_mark: | `-0.17%` :no_entry:   | `0` :heavy_check_mark: |


The following classes were added in #1760:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.ElasticsearchV7Constants](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `1.00%` | `1` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.Transport_Request_Integration](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `0.70%` | `11` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.Transport_RequestAsync_Integration](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `0.70%` | `11` |


1 classes were removed from Datadog.Trace in #1760


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75197/artifacts?artifactName=Code%20Coverage%20Report_75197&api-version=6.0&%24format=zip)
* [For this PR #1760](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75180/artifacts?artifactName=Code%20Coverage%20Report_75180&api-version=6.0&%24format=zip)

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will **not change** line coverage
:heavy_check_mark: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will **not change** branch coverage
:no_entry: Merging #1760 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) will will **increase** complexity by `25`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) | #1760       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9314` / `14757` | `9326` / `14772` |          |
| Lines %   | `0.63%`      | `0.63%`      |  `0.00%` :heavy_check_mark:  |
| Branches  | `4126` / `6418` | `4141` / `6440` |          |
| Branches %| `0.64%`      | `0.64%`      |  `0.00%` :heavy_check_mark:  |
| Complexity|   `7558`      | `7583`        |  `25`  :no_entry:    |

View the full report for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75197/artifacts?artifactName=Code%20Coverage%20Report_75197&api-version=6.0&%24format=zip)
* [For this PR #1760](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75349/artifacts?artifactName=Code%20Coverage%20Report_75349&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/16dd0fce121ca0fe9b20e650c05823496d603283) | #1760       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `0.63%`      | `0.63%`       |  `0.00%` :heavy_check_mark:  |
| Branches %| `0.64%`      | `0.64%`       |  `0.00%` :heavy_check_mark:  |
| Complexity| `7558`      | `7583`       |  `25` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations.TaskContinuationGenerator`4](https://github.com/DataDog/dd-trace-dotnet/tree/1215c88a5610e6f8cbfd6cdb93fa6b7a60ae6b2b/tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/Continuations/TaskContinuationGenerator`1.cs) | `0.05%` :heavy_check_mark: | `0.09%` :heavy_check_mark:   | `0` :heavy_check_mark: |


The following classes were added in #1760:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.ElasticsearchV7Constants](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `1.00%` | `1` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.Transport_Request_Integration](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `0.70%` | `11` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Elasticsearch.V7.Transport_RequestAsync_Integration](https://github.com/DataDog/dd-trace-dotnet/pull/1760/files) | `1.00%` | `0.70%` | `11` |


View the full reports for further details:

* [For master](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75197/artifacts?artifactName=Code%20Coverage%20Report_75197&api-version=6.0&%24format=zip)
* [For this PR #1760](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75349/artifacts?artifactName=Code%20Coverage%20Report_75349&api-version=6.0&%24format=zip)

**Pull request merged**